runtime/compiler/codegen/J9AheadOfTimeCompile.cpp

```diff
@@ -546,12 +546,27 @@ J9::AheadOfTimeCompile::initializeCommonAOTRelocationHeader(TR::IteratedExternal
          if (aconstNode->getOpCodeValue() == TR::loadaddr)
             j9method = reinterpret_cast<TR_OpaqueMethodBlock *>(aconstNode->getSymbolReference()->getSymbol()->castToStaticSymbol()->getStaticAddress());
 
-         TR_OpaqueClassBlock *j9class = fej9->getClassFromMethodBlock(j9method);
+         TR_InlinedCallSite & ics = comp->getInlinedCallSite(inlinedSiteIndex);
+         TR_ResolvedMethod *inlinedMethod = ((TR_AOTMethodInfo *)ics._methodInfo)->resolvedMethod;
+         TR_OpaqueMethodBlock *inlinedJ9Method = inlinedMethod->getPersistentIdentifier();
 
-         uintptr_t classChainOffsetOfCLInSharedCache = sharedCache->getClassChainOffsetOfIdentifyingLoaderForClazzInSharedCache(j9class);
-         uintptr_t classChainForInlinedMethodOffsetInSharedCache = self()->getClassChainOffset(j9class);
+         uintptr_t classChainOffsetOfCLInSharedCache = 0;
+         uintptr_t classChainForInlinedMethodOffsetInSharedCache = 0;
+         uintptr_t vTableOffset = 0;
 
-         uintptr_t vTableOffset = static_cast<uintptr_t>(fej9->getInterpreterVTableSlot(j9method, j9class));
+         /* If the j9method from the aconst node is the same as the j9method at
+          * inlined call site at inlinedSiteIndex, don't set the remaining fields
+          * in the binary template; at relo time, the inlined site index is
+          * sufficient to materialize the j9method pointer
+          */
+         if (j9method != inlinedJ9Method)
```

Comment:
imo a better way to do this would be to create a new relocation record that does the inlining relocation as you want it, rather than using the existing `MethodPointer` relocation. That would avoid storing empty payload data as a flag to "process this one differently".

Reply:
I think that's in general a good idea. However, for the purposes of this PR, it might be best done in a separate PR later:

1. Adding a new relo type requires a change to OMR, so that's tedious; additionally there are uses of `TR_MethodPointer` in OMR so we would need to do a coordinate change where we're not only adding FSD but also modifying unrelated relo changes. Or we'd have to do that in a separate PR prior to these changes, in which case we might as well do it after.
2. `TR_MethodPointer` does nicely encapsulate the idea of "we have a pointer to a j9method". As such,  to determine in the code when we would use `TR_MethodPointer` and when we'd use the other might require us to check the inlinedSiteIndex every time we went to add the relo. This might be fine, but not something I want to do in this PR, especially because it should be done in a way that people shouldn't have to remember to always add the if.

Personally I don't mind the current approach ONLY because the "if this then that" logic is hidden away, and the two branches are still related to the same relocation. However, I do agree that it would be nice to have logical separation. So for the sake of not slowing down this PR, I'd prefer to just add it later on; in fact it could probably be bundled with the clean up that's needed in some of the inlined method relos where we use the cpIndex field for the SVM IDs heh.

Reply:
ok, I accept that the better approach would be a more invasive change. But I would prefer that better change. It might be possible to handle it transparently by having the MethodPointer external iterated relocation constructor automatically morph itself into a different relocation kind if it detects that the pointer matches the inlined site's j9method point. That doesn't change the fact that you would need an OMR change as well...

Reply:
if only the relocation framework were extensible....